File: src/generators/templates/javascript/file.js
// === Sandbox Executor JavaScript Proxy (File Mode for QuickJS) ===
import * as std from 'std';

const START_MARKER = "__SANDBOX_RESULT_START__";
const END_MARKER = "__SANDBOX_RESULT_END__";

async function executeRequest(request) {
  try {
    const filePath = request.filePath || './user_code.js';
    const funcName = request.functionName;
    const args = request.args || [];
    const kwargs = request.kwargs || {};

    const userModule = await import(filePath);
    const func = userModule[funcName];
    if (typeof func !== 'function') {
      throw new Error(`Function '${funcName}' not found in ${filePath}`);
    }

    let result;
    if (Object.keys(kwargs).length > 0) {
      result = await func(...args, kwargs);
    } else {
      result = await func(...args);
    }

    return {
      success: true,
      result: result
    };
  } catch (e) {
    return {
      success: false,
      error: {
        message: e.message || String(e),
        type: e.name || 'Error',
        stack: e.stack
      }
    };
  }
}

async function main() {
  try {
    const requestFile = '/workspace/.sandbox_request.json';
    const content = std.loadFile(requestFile);
    if (!content) {
      throw new Error(`Request file not found: ${requestFile}`);
    }

    const request = JSON.parse(content);
    const output = await executeRequest(request);
    
    console.log(START_MARKER);
    console.log(JSON.stringify(output));
    console.log(END_MARKER);
  } catch (err) {
    console.log(START_MARKER);
    console.log(JSON.stringify({
      success: false,
      error: { message: "Fatal: " + (err.message || String(err)), type: 'FatalError' }
    }));
    console.log(END_MARKER);
  }
}

main();

File: src/generators/templates/ruby/file.rb
# === Sandbox Executor Ruby Proxy (File Mode) ===
require 'json'

START_MARKER = "__SANDBOX_RESULT_START__"
END_MARKER = "__SANDBOX_RESULT_END__"

def execute_request(request)
  begin
    file_path = request['filePath'] || '/workspace/user_code.rb'
    func_name = request['functionName']
    args = request['args'] || []
    kwargs = request['kwargs'] || {}

    require_relative file_path.sub(/\.rb$/, '')

    if kwargs && !kwargs.empty?
      symbol_kwargs = kwargs.transform_keys(&:to_sym)
      result = send(func_name, *args, **symbol_kwargs)
    else
      result = send(func_name, *args)
    end

    {
      "success" => true,
      "result" => result
    }
  rescue => e
    {
      "success" => false,
      "error" => {
        "message" => e.message,
        "type" => e.class.name,
        "stack" => e.backtrace.join("
")
      }
    }
  end
end

begin
  request_file = '/workspace/.sandbox_request.json'
  if File.exist?(request_file)
    payload = File.read(request_file)
    request = JSON.parse(payload)
    output = execute_request(request)
    puts START_MARKER
    puts JSON.generate(output)
    puts END_MARKER
  else
    raise "Request file not found: #{request_file}"
  end
rescue => e
  output = {
    "success" => false,
    "error" => {
      "message" => "Fatal: " + e.message,
      "type" => e.class.name,
      "stack" => e.backtrace&.join("
")
    }
  }
  puts START_MARKER
  puts JSON.generate(output)
  puts END_MARKER
end
